src/looprestoration.rs

```diff
@@ -1556,32 +1556,1845 @@ unsafe fn sgr_filter_3x3_neon<BD: BitDepth>(
+    }
+
+    let mut sumsq5_ptrs: [*mut i32; 5] = [0 as *mut i32; 5];
+    let mut sum5_ptrs: [*mut i16; 5] = [0 as *mut i16; 5];
+    for i in 0..5 {
+        sumsq5_ptrs[i] = sumsq5_rows[if lr_have_top && i > 0 { i - 1 } else { 0 }];
+        sum5_ptrs[i] = sum5_rows[if lr_have_top && i > 0 { i - 1 } else { 0 }];
+    }
+
+    if lr_have_top {
+        rav1d_sgr_box35_row_h_neon(
+            sumsq3_rows[0],
+            sum3_rows[0],
+            sumsq5_rows[0],
+            sum5_rows[0],
+            0 as *const LeftPixelRow<BD::Pixel>,
+            lpf,
+            w,
+            edges,
+            bd,
+        );
+        lpf = lpf.add(stride);
+        rav1d_sgr_box35_row_h_neon(
+            sumsq3_rows[1],
+            sum3_rows[1],
+            sumsq5_rows[1],
+            sum5_rows[1],
+            0 as *const LeftPixelRow<BD::Pixel>,
+            lpf,
+            w,
+            edges,
+            bd,
+        );
+
+        rav1d_sgr_box35_row_h_neon(
+            sumsq3_rows[2],
+            sum3_rows[2],
+            sumsq5_rows[2],
+            sum5_rows[2],
+            left,
+            src,
+            w,
+            edges,
+            bd,
+        );
+
+        left = left.offset(1);
+        src = src.add(stride);
+
+        sgr_box3_vert_neon(
+            &mut sumsq3_ptrs,
+            &mut sum3_ptrs,
+            A3_ptrs[3],
+            B3_ptrs[3],
+            w,
+            (*params).sgr.s1 as c_int,
+            bd,
+        );
+        rotate_ab_4(&mut A3_ptrs, &mut B3_ptrs);
+
+        h -= 1;
+        if h <= 0 {
+            track = Track::vert1;
+        } else {
+            rav1d_sgr_box35_row_h_neon(
+                sumsq3_ptrs[2],
+                sum3_ptrs[2],
+                sumsq5_rows[3],
+                sum5_rows[3],
+                left,
+                src,
+                w,
+                edges,
+                bd,
+            );
+            left = left.offset(1);
+            src = src.add(stride);
+
+            sgr_box5_vert_neon(
+                &mut sumsq5_ptrs,
+                &mut sum5_ptrs,
+                A5_ptrs[1],
+                B5_ptrs[1],
+                w,
+                (*params).sgr.s0 as c_int,
+                bd,
+            );
+            rotate_ab_2(&mut A5_ptrs, &mut B5_ptrs);
+
+            sgr_box3_vert_neon(
+                &mut sumsq3_ptrs,
+                &mut sum3_ptrs,
+                A3_ptrs[3],
+                B3_ptrs[3],
+                w,
+                (*params).sgr.s1 as c_int,
+                bd,
+            );
+            rotate_ab_4(&mut A3_ptrs, &mut B3_ptrs);
+
+            h -= 1;
+            if h <= 0 {
+                track = Track::vert2;
+            }
+
+            // ptrs are rotated by 2; both [3] and [4] now point at rows[0]; set
+            // one of them to point at the previously unused rows[4].
+            sumsq5_ptrs[3] = sumsq5_rows[4];
+            sum5_ptrs[3] = sum5_rows[4];
```

Comment:
Should these lines be in an `else` on the previous `if`? That's the structure you've been using to translate the `goto`s, and these lines were after the `goto` in the original example.

Reply:
Good catch! I fixed it.